cmd/sync/sync.go

```diff
@@ -3,64 +3,232 @@
+				if e := mq.SendJSONError(&delivered, err.Error(), conf.Broker); e != nil {
+					log.Error("failed to publish message, reason: ", res.Errors())
+				}
+				// Restart on new message
+				continue
+			}
+
+			// we unmarshal the message in the validation step so this is safe to do
+			_ = json.Unmarshal(delivered.Body, &message)
+
+			// Extract the sha256 from the message and use it for the database
+			var checksumSha256 string
+			for _, checksum := range message.DecryptedChecksums {
+				if checksum.Type == "sha256" {
+					checksumSha256 = checksum.Value
+				}
+			}
+
+			c := completed{
+				User:               message.User,
+				Filepath:           message.Filepath,
+				AccessionID:        message.AccessionID,
+				DecryptedChecksums: message.DecryptedChecksums,
+			}
+
+			completeMsg, _ := json.Marshal(&c)
+
+			res, err = validateJSON(conf.SchemasPath, completeMsg)
+			if err != nil {
+				log.Errorf("json error: %v", err)
+				// Nack message so the server gets notified that something is wrong but don't requeue the message
+				if e := delivered.Nack(false, false); e != nil {
+					log.Errorln("failed to Nack message, reason: ", e)
+				}
+				// Send the message to an error queue so it can be analyzed.
+				if e := mq.SendJSONError(&delivered, err.Error(), conf.Broker); e != nil {
+					log.Error("failed to publish message, reason: ", err)
+				}
+				// Restart on new message
+				continue
+			}
+			if !res.Valid() {
+				log.Errorf("result.error: %v", res.Errors())
+				log.Error("Validation failed")
+				// Nack message so the server gets notified that something is wrong but don't requeue the message
+				if e := delivered.Nack(false, false); e != nil {
+					log.Errorln("failed to Nack message, reason: ", e)
+				}
+				// Send the message to an error queue so it can be analyzed.
+				if e := mq.SendJSONError(&delivered, err.Error(), conf.Broker); e != nil {
+					log.Error("failed to publish message, reason: ", err)
+				}
+				// Restart on new message
+				continue
+			}
+
+			log.Debug("Sync initiated")
+			file, err := inbox.NewFileReader(message.Filepath)
+			if err != nil {
+				log.Errorf("Failed to open file: %s, reason: %v", message.Filepath, err)
+				continue
+			}
+
+			dest, err := backup.NewFileWriter(message.Filepath)
+			if err != nil {
+				log.Errorf("Failed to create file, reason: %v", err)
+				continue
+			}
+
+			_, err = io.Copy(dest, file)
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			file.Close()
+			dest.Close()
+			log.Debugln("Sync completed")
+
+			log.Debug("Mark ready")
+			if err := db.MarkReady(message.AccessionID, message.User, message.Filepath, checksumSha256); err != nil {
+				log.Errorf("MarkReady failed, reason: %v", err)
+				// nack the message but requeue until we fixed the SQL retry.
+				if e := delivered.Nack(false, true); e != nil {
+					log.Errorln("failed to Nack message, reason: ", e)
+				}
+				continue
+				// this should be handled by the SQL retry mechanism
+			}
```

Comment:
This belongs to finalize?

Reply:
The idea is to deploy either sync or finalize. So I just included the sync in the already existing code of finalize

Reply:
Ah, that wasn't my understanding but in that case I guess it's fine.